### PR TITLE
Categorical covariate for regression models

### PR DESCRIPTION
**Pull request by jonasblanc**
Checklist before merging this PR:
- [x] Mentioned all issues that this PR fixes or addresses.
- [x] Summarized the updates of this PR under **Summary**.
- [x] Added an entry under **Unreleased** in the [Changelog](../CHANGELOG.md).

Fixes #.

Partially addresses #1920 and #1514 
Necessity for categorical covariate focused PR emerges from #2737 

### Summary
- Rewrite the categorical covariate matching logic in `RegressionModelWithCategoricalCovariates` (between components names in `init` and `fit`/`predict` input data).
- Refactor logic to support models specific way of flagging input features as categorical.
- Add support for categorical covariate in `CatBoostModel`.
- Make `RegressionModelWithCategoricalCovariates` abstract, define function for the subclass to override for model specific format/behavior.
- Explore categorical covariate support for `XGBoostModel` and `HistGradientBoostingRegressor`. (It was decided to limit categorical covariate support for `CatBoostModel` and `LightGBM`. If this decision was to change in the future, please see [920ff40](https://github.com/unit8co/darts/pull/2750/commits/920ff40747da975d61785eab118c7495c6634306) for implementation of categorical covariate support for `XGBoostModel` and `HistGradientBoostingRegressor` )
- Extend tests to cover the broader categorical covariate support.

### Other Information

**How models treat categorical features**
(This is based on my current understanding, any rectification is welcome)

- `CatBoostModel`: [Categorical features](https://catboost.ai/docs/en/features/categorical-features)
    - Categorical features are re-encoded (sometime combined) into float (without relying on label value)
    - Many parameters to tune categorical support notably `one_hot_max_size`

- `LightGBM`: [Categorical Feature Support](https://lightgbm.readthedocs.io/en/latest/Advanced-Topics.html#categorical-feature-support) / [categorical_feature](https://lightgbm.readthedocs.io/en/latest/Parameters.html#categorical_feature)
    - (integer codes will be extracted from pandas categoricals in the Python-package) we are not using pandas
    - Categorical features are cast to `int32` (must be non-negative value, ideally [0, 1, .., C-1])
    - according to [a kaggle forum](https://www.kaggle.com/code/mlisovyi/beware-of-categorical-features-in-lgbm/comments), since the features will be discretized into max_bin the encoding of categorical features won't matter unless the cardinality of the feature is larger than max_bin, in which case the encoding will dictate the grouping. (Similarly to `HistGradientBoostingRegressor`).
    - applies [Fisher (1958)](https://www.tandfonline.com/doi/abs/10.1080/01621459.1958.10501479) to find the optimal split over categories
    - for high cardinality best to treat feature as numeric (either by ignoring categorical interpretation or by embedding features in low-dim space

- `HistGradientBoostingRegressor `: [1.11.1.1.4. Categorical Features Support](https://scikit-learn.org/stable/modules/ensemble.html#categorical-features-support) / [categorical_features](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.HistGradientBoostingRegressor.html#sklearn.ensemble.HistGradientBoostingRegressor:~:text=larger%20than%20255.-,categorical_features) / [Example: Categorical Feature Support in Gradient Boosting](https://scikit-learn.org/stable/auto_examples/ensemble/plot_gradient_boosting_categorical.html) / [Examples: Comparing Target Encoder with Other Encoders](https://scikit-learn.org/stable/auto_examples/preprocessing/plot_target_encoder.html#sphx-glr-auto-examples-preprocessing-plot-target-encoder-py)
  - Re-encode categorical feature with OrdinalEncoder
  - Re-encoded categorical features considered as ordered numerical, group them by bucket based on their value
  - Possibility to specify cardinality of category under which to use one hot encoding: `max_cat_to_onehot`

 - `XGBoost`: [Tuto Categorical Data](https://xgboost.readthedocs.io/en/latest/tutorials/categorical.html) / [Parameters for categorical features](https://xgboost.readthedocs.io/en/stable/parameter.html#parameters-for-categorical-feature) / [Examples categorical features](https://xgboost.readthedocs.io/en/stable/python/examples/cat_in_the_dat.html#sphx-glr-python-examples-cat-in-the-dat-py) / 
 [Examples categorical pipeline](https://xgboost.readthedocs.io/en/stable/python/examples/cat_pipeline.html#sphx-glr-python-examples-cat-pipeline-py)
   - Re-encode categorical features through pandas cat.codes (issue with persistence, fix on the way)
   - Re-encoded categorical features considered as ordered numerical ??
     - Partition split ?

**Decision to limit categorical covariate support to `LightGBM` and `CatBoostModel`.**
- `CatBoostModel` is meant to deal with categorical features natively thus adding support for this model makes perfect sense.
- `XGBoost` is still in the process of making their encoding consistent, please see [920ff40](https://github.com/unit8co/darts/pull/2750/commits/920ff40747da975d61785eab118c7495c6634306) if you would like to add support on your side.
- `RegressionModelWithCategoricalCovariates` is meant to be an abstract class. `HistGradientBoostingRegressor` categorical native support does not add much plus-value when categories are integer already, we let it to the user to extend `RegressionModelWithCategoricalCovariates` to support it (please see [920ff40](https://github.com/unit8co/darts/pull/2750/commits/920ff40747da975d61785eab118c7495c6634306) for an example of how to support categorical covariate in `HistGradientBoostingRegressor`).

**Discussion about categorical support for `XGBoostModel`.**
When specifying a component is categorical through category type of pandas dataframe, XGB re-encodes the columns through pandas .cat.codes method. The encoding is not consistent if not all categories are present at encoding time, which is an issue as we aim at keeping darts' model and encoder separated. 

There is a current [effort](https://github.com/dmlc/xgboost/issues/11088) for auto-encoding categorical data during inference to eliminate the need for a pipeline for consistent encoding beforehand.   

It's still unclear to me if categorical features of subtype `int` necessarily under-go a re-encoding or if we could simply assume our categorical covariates are valid encodings (and enforce it) and flag it as categorical without xgb re-encoding it.

**Discussion about categorical support for `HistGradientBoostingRegressor`.**
`HistGradientBoostingRegressor` is the only sklearn model that is supposed to support categorical features natively. [In practice](https://github.com/scikit-learn/scikit-learn/blob/98ed9dc73a86f5f11781a0e21f24c8f47979ec67/sklearn/ensemble/_hist_gradient_boosting/gradient_boosting.py#L278C9-L278C11) it is using an `OrdinalEncoder`on categorical features internally. Whether it makes sense in our specific case where all input are numerical is debatable. 

**Converting samples np.array to Pandas dataframe.**
Both `CatBoost` and `XGBoostModel` require features having different data types depending on if they are categorical or not. We convert input samples to `pandas dataframe` to type categorical features which is not ideal for large input. However multiple `np.array` each one containing a typed feature is not an option as a list of `np.array` will be interpreted as a list of sample and not features per the models.

### CODE REVIEW

File: darts/models/forecasting/regression_model.py [hunk 1]
```diff
@@ -2048,3 +2022,40 @@ def _fit_model(
             sample_weight=sample_weight,
             **kwargs,
         )
+
+    def _format_samples(self, samples, labels=None):
```
Comment (dennisbader) [inline darts/models/forecasting/regression_model.py:2026]: this can be removed since we drop HistGradientBoostingRegressor. It should probably be an abstractmethod

File: darts/tests/models/forecasting/test_regression_models.py [hunk 2]
```diff
@@ -403,15 +396,20 @@ def _apply_promo_mechanism(promo_mechanism):
...
         date_range = pd.date_range(start="2020-01-01", end="2023-01-01", freq="D")
         df = (
             pd.DataFrame({
                 "date": date_range,
                 "baseline": np.random.normal(100, 10, len(date_range)),
                 "fut_cov_promo_mechanism": np.random.randint(0, 11, len(date_range)),
                 "fut_cov_dummy": np.random.normal(10, 2, len(date_range)),
-                "past_cov_dummy": np.random.normal(10, 2, len(date_range)),
-                "past_cov_cat_dummy": np.random.normal(10, 2, len(date_range)),
+                "past_cov_dummy_neg_int": np.array([
```
Comment (dennisbader) [inline darts/tests/models/forecasting/test_regression_models.py:407]: this neg int can probably removed once the validation check for it was removed, right?

**Review by dennisbader — CHANGES_REQUESTED** [review]
Thanks a lot for this @jonasblanc, looks already very nice 🚀 

I left some suggestions where I believe we could reduce code and simplify a bit.

I couldn't yet fully go through the tests, but will do that in the next review round :)

File: darts/models/forecasting/regression_model.py [hunk 3]
```diff
@@ -682,6 +684,14 @@ def _create_lagged_data(
 
         return features, labels, sample_weights
 
+    def _format_samples(self, samples, labels=None):
```
[the remark this answers is not in the document]
Comment (jonasblanc) [reply darts/models/forecasting/regression_model.py:687]: I'll add type annotations for arguments but we cannot predict type of output as it depends on the format required by the model.

File: darts/models/forecasting/regression_model.py [hunk 4]
```diff
@@ -1956,64 +1905,77 @@ def _get_categorical_features(
             in create_lagged_data.
         2. Get the indices of the categorical features in the list of features.
         """
+        categorical_covariates = [
```
[the remark this answers is not in the document]
Comment (jonasblanc) [reply darts/models/forecasting/regression_model.py:1908]: They are (maybe not in this form but rewiting them is necessary). 
The previous version of `_get_categorical_features` had several vulnerabilities:
- name matching was performed across all covariates meaning that if two composents of differents covariates would have the same name matching one of the categorical covariate name then both would be flagged as categorical.
- the test condition for matching composent was done as a substring inclusion condition, meaning that if one composent had a name that would include the name of an other categorical component, then again both would be flagged as categorical.

`_validate_categorical_covariates` on the other hand was responsible for not supporting a categorical covariate that would be created by `add_encoders`. Indeed it would raise an error since the encoders new component would not be created at the time of `_validate_categorical_covariates`. That's why it was required to move the checks to `_get_categorical_features` which makes sure are specified categorical covariates have a matching.

File: darts/models/forecasting/regression_model.py [hunk 5]
```diff
@@ -2029,17 +1991,29 @@ def _fit_model(
...
 
+        # cat_param_name is None if model has no attribute to specify categorical features in `fit()`
+        # cat_param_default is None if no default value to auto-detect categorical features is available in `fit()`
         cat_param_name, cat_param_default = self._categorical_fit_param
-        kwargs[cat_param_name] = (
-            cat_col_indices if cat_col_indices else cat_param_default
-        )
+
+        # Add attributes in kwargs if applicable for model's `fit()` method
+        if cat_param_name is not None and (
```
[the remark this answers is not in the document]
Comment (jonasblanc) [reply darts/models/forecasting/regression_model.py:2010]: I believe the reason you wrote
````
kwargs[cat_param_name] = (
            cat_col_indices if cat_col_indices else cat_param_default
        )
`````
2 years ago was that LightGBM offers to automatically infer the categorical features if `categorical_feature` is set to "auto". So I adapted it for catboost as well.
After more digging it turns out the "auto" parameter is only relevant when fit receives pandas df wich we are not using for LightGBM. So I guess you're right and we can remove cat_param_default altogether.

**Review by dennisbader — APPROVED** [review]
This looks great, thanks a lot @jonasblanc 🚀 

I made some minor changes, mainly in the tests (one new test that checks runnability of models using cat features and likelihoods), and some other small refactors.

We can merge after all tests have passed 💯
